### PR TITLE
Default grace period to 0 when --force is used to delete an object

### PR DESCRIPTION
**Pull request by brianpursley**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
When deleting objects, kubectl currently requires you to specify `--grace-period=0` when you use `--force`.  There are two usability problems with this:

1.  This is redundant because `--grace-period=0` is only allowed to be specified if you also pass `--force`, enforced by a validation check.

2.  If you forget to provide `--grace-period=0` when using `--force`, you get a warning telling you that you should have used `--grace-period=0` and then the delete operation still proceeds proceeds, but without force.  This can be sort of frustrating since you clearly specified `--force` so your intention to force was already indicated, yet kubectl proceeded in a non-force way.

To fix these two problems, this PR eliminates the need to specify `--grace-period=0` when using `--force` by defaulting grace period to 0 when force is used.

Because I discovered there are several different validation checks performed on various combinations of `--force`, `--grace-period` and `--now`, I also added unit test cases for each of these scenarios to help make sure the expected behavior is clear and tested for during unit testing in the future.

**Which issue(s) this PR fixes**:

https://github.com/kubernetes/kubectl/issues/813

**Special notes for your reviewer**:
None

**Does this PR introduce a user-facing change?**:

```release-note
When deleting objects using kubectl with the --force flag, you are no longer required to also specify --grace-period=0.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
None

### CODE REVIEW

**Comment by brianpursley** [comment]
/cc @mengqiy
Looking for ok to test...

**Comment by brianpursley** [comment]
Squashed commits

**Comment by neolit123** [comment]
@kubernetes/sig-cli-pr-reviews

File: staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go]: This nested ifs complicate the flow a lot and mixes filling the internal structure (`Complete` method) with validation (`Validate` method), I'd suggest just add:
```go
if o.ForceDeletion && o.GracePeriod <= 0 {
    o.GracePeriod = 0
}
```
after the old conditions.

File: staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go]: Add here validation:
```go
if o.GracePeriod > 0 && o.ForceDeletion {
    return fmt.Errof("--force and --grace-period greater than 0 cannot be specified together")
}
```

**Review by soltysh — CHANGES_REQUESTED** [review]
Please ping me on slack when you get this updated (same nick :wink: )

File: staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go]: Good idea, much cleaner, thanks.

File: staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go]: Agreed.  This is a much cleaner change.  Thanks

File: staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go]: Agreed. This is a much cleaner change. Thanks

**Review by soltysh — APPROVED** [review]
/approve
You still need to fix the bazel, run `hack/update-bazel.sh` and you're good to go

**Comment by brianpursley** [comment]
@soltysh ptal, needs lgtm for tide

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by brianpursley** [comment]
/hold

OK it looks like there is some kind of problem blocking the rebuild from succeeding:
```
ERROR(all) vendor/k8s.io/kubectl/pkg/cmd/delete/delete_test.go:38:2: cmdutil redeclared in this block
ERROR(all) vendor/k8s.io/kubectl/pkg/cmd/delete/delete_test.go:23:2: 	other declaration of cmdutil
exit status 1
!!! Type Check has failed. This may cause cross platform build failures.
!!! Please see https://git.k8s.io/kubernetes/test/typecheck for more information.
```

I'll investigate later today or tomorrow.

**Comment by brianpursley** [comment]
Tried to reproduce locally by running `hack/verify-typecheck-providerless.sh`.
I rebased and will see if the test passes now.

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**
